Feature/dev docker

st3v3nhunt opened the pull request:
To have code running locally the changes in https://github.com/nhsuk/connecting-to-services/pull/239 are required

For the API this branch is required https://github.com/nhsuk/nearby-services-api/tree/feature/0.2.1-beta

adamgraves-choices committed:
Add docker-compose file for local development

st3v3nhunt committed:
:heart: Use env file for default env var values

st3v3nhunt committed:
:ram: Build local projects for development

st3v3nhunt committed:
:chestnut: Better instructions for setting up dev env

st3v3nhunt commented:
![](https://github.trello.services/images/mini-trello-icon.png) [All developers able to make code changes and test them using docker.](https://trello.com/c/lXDzR8a0/732-all-developers-able-to-make-code-changes-and-test-them-using-docker)

st3v3nhunt committed:
:door: Use env var for code dir

st3v3nhunt committed:
:bell: Add services to run tests

beseven commented:
@st3v3nhunt - myself and @markysoft hit the issue that if you haven't done `npm install` locally / on the api - the docker image will not work. There is a fix in https://github.com/nhsuk/nhsuk-rancher-templates/blob/feature/live-copy-dev-docker/templates/c2s_local_dev/docker-compose.yml#L39 for that as the expectation might be to not have to do that.  On another note - given we've branched off the 0.2.0-beta tag, a similar Docker file is needed in the nearby-service-api as its needed for the front-end app.

st3v3nhunt committed:
:gun: Parameterise NODE_ENV for conditional RUN

st3v3nhunt committed:
:bath: Use code rather than app

neilbmclaughlin commented on templates/c2s_local_dev/docker-compose.yml:
```diff
@@ -0,0 +1,65 @@
+    environment:
+      NODE_ENV: ${NODE_ENV}
+      SPLUNK_HEC_ENDPOINT: ${SPLUNK_HEC_ENDPOINT}
+      SPLUNK_HEC_TOKEN: ${SPLUNK_HEC_TOKEN}
```
Works as is except, when running against master of the api, then needed to add `DB_PRIMARY_KEY: ${DB_PRIMARY_KEY}` to the environment section of the web-app-test and nearby-services-api service. Also require the actual key in .env

Just noting it explicitly here for others. The requirement to update .env/docker-compose to add any other variables is covered in the README.

neilbmclaughlin commented on templates/c2s_local_dev/docker-compose.yml:
```diff
@@ -0,0 +1,65 @@
+    build:
+      context: ${CODE_DIR}/connecting-to-services/
+      args:
+        NODE_ENV: ${NODE_ENV}
```
@beseven raised the question should NODE_ENV always be 'test' in this case.

st3v3nhunt replied:
Good point. I'll make the change for both test services.

neilbmclaughlin commented on templates/c2s_local_dev/docker-compose.yml:
```diff
@@ -0,0 +1,65 @@
+    build:
+      context: ${CODE_DIR}/nearby-services-api/
+      args:
+        NODE_ENV: ${NODE_ENV}
```
See above regarding NODE_ENV

neilbmclaughlin commented on templates/c2s_local_dev/docker-compose.yml:
```diff
@@ -0,0 +1,65 @@
+      - api-deps:/api/node_modules
+
+volumes:
+  app-deps:
```
@adamgraves-choices questioned whether sharing volumes is an issue.

neilbmclaughlin commented:
Will take a look tomorrow when running against the two specific branches mentioned above.

Would it be possible to create a PR for the feature/local-dev branch to make it easier to see the differences.

neilbmclaughlin commented:
Running `docker-compose up --build --force-recreate` for the two feature\local-dev branches above gives the following error:

```
api-test_1             | > NODE_ENV=test mocha --recursive test "--watch" "--reporter" "min"
api-test_1             |
api-test_1             | Error: Cannot find module 'documentdb'
api-test_1             |     at Function.Module._resolveFilename (module.js:470:15)
api-test_1             |     at Function.Module._load (module.js:418:25)
<snip>
```

Opening a shell on api-test_1 and running 

```
docker exec -it c2slocaldev_api-test_1 sh
/code # npm test
```

gives the same error.

If I run `npm install` on the host then running `npm test` in the container gets further (missing variable exception) but running `npm install && npm test` on the container results in the documentdb error again. There is odd stuff going on so will investigate more tomorrow.

st3v3nhunt commented:
@neilbmclaughlin The errors might be due to me having rebased `feature/local-dev` for the api against master which brought in the documentDB changes. I didn't mean to do this.

Can you try it whilst having the `feature/0.2.1-beta` branch checked out for the api (https://github.com/nhsuk/nearby-services-api/tree/feature/0.2.1-beta)? Even if there are still problems the stuff with docDB can be ruled out.

st3v3nhunt committed:
:mouse: Use test NODE_ENV for test services

st3v3nhunt committed:
:foggy: Do not share volumes between services

It causes problems as the different services are writing/updating the
volume at different times and it becomes an unatomic operation.

st3v3nhunt commented:
The last commit (https://github.com/nhsuk/nhsuk-rancher-templates/pull/2/commits/dc4b42d022aecb27a5a646046d443c0e267d81b4) fixes the issue of inconsistently missing packages in the running services.

st3v3nhunt commented:
Superseded by #7
